### PR TITLE
Added documentation for e2e tests

### PR DESCRIPTION
**Pull request by ShaunakJoshi1407**
This commit adds the documentation for e2e tests. Fixes : https://github.com/arlonproj/arlon/issues/279

### CODE REVIEW

File: testing/e2e_docs.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+## Arlon e2e tests
+
+
+### Setup
+
+- First, we run the ```testing/e2e_setup.sh``` script, which helps us setup a kind management cluster, a git-server [gitea](https://gitea.io/en-us) based workspace repository and installs services like argocd, arlon, capi in the management cluster. It also installs other required tools like kubectl, docker, kind, kuttl, clusterctl, helm, gitea.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:6]:
```suggestion
- First, we run the `testing/e2e_setup.sh` script, which helps us setup a kind management cluster, a git-server [gitea](https://gitea.io/en-us) based workspace repository and installs services like argocd, arlon, capi in the management cluster. It also installs other required tools like kubectl, docker, kind, kuttl, clusterctl, helm and gitea.
```

File: testing/e2e_docs.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  
+- In addition to this, this script also creates a capi-eks cluster manifest which serves as the manifest for the workload cluster. This cluster is created as a part of the e2e tests.This manifest is pushed into the workspace repository created in the previous step
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:8]:
```suggestion
- In addition to this, this script also creates a capi-eks cluster manifest which serves as the basecluster manifest. This cluster is created as a part of the e2e tests and is pushed to the workspace repository created in the previous step.
```

File: testing/e2e_docs.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  
+- This script also adds the xenial bundle manifests to the workspace repository. This is required for creating a xenial bundle and a corresponding profile during the cluster update operation in the e2e tests.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:10]:
```suggestion
- This script also adds the xenial bundle manifest to the workspace repository which is required for creating a xenial bundle and a corresponding profile which is consumed by the test.
```

File: testing/e2e_docs.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+- We are using [KUTTL](https://kuttl.dev/) to write e2e integration tests for arlon.
+
+- KUTTL is a declarative integration testing harness for testing operators, KUDO, Helm charts, and any other Kubernetes applications or controllers.
+
+- The KUTTL test CLI organizes tests into suites:
+  - A "test step" defines a set of Kubernetes manifests to apply and a state to assert on (wait for or expect).
+  - A "test assert" are the part of a test step that define the state to wait for Kubernetes to reach
+  
+- All the e2e tests are placed in ```/testing/e2e``` in the respecitive directory for the given test. For example, the test 00-deploy is used to deploy a cluster and all the files related to this test are placed in ```/testing/e2e/00-deploy```. To add a new e2e test here, create a new directory with the correct step index. The tests in ```/testing/e2e``` run in the order of the index specified.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:25]:
```suggestion
- All the e2e tests are placed in `/testing/e2e` in the respecitive directory for the given test. For example, the test `00-deploy` is used to deploy a cluster and all the files related to this test are placed in `/testing/e2e/00-deploy`. To add a new e2e test here, create a new directory with the correct step index. The tests in `/testing/e2e` run in the order of the index specified.
```

File: testing/e2e_docs.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+- Each filename in the test case directory should start with an index (in this example 00) that indicates which test step the file is a part of. The first filename in the test case will begin with index 00, the second will have index 01 and so on. e.g In ```/testing/e2e/00-deploy``` we have 00-prepare.yaml, 01-validate.yaml etc.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:27]:
```suggestion
- Each filename in the test case directory should start with an index (in this example 00) that indicates which test step the file is a part of. The first filename in the test case will begin with index 00, the second will have index 01 and so on. e.g In `/testing/e2e/00-deploy` we have `00-prepare.yaml`, `01-validate.yaml` etc.
```

File: testing/e2e_docs.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+- Do note that the tests will run in the order of the index specified. This is the case for both the test case directory and the individual test files within these directories. Files that do not start with a step index are ignored and can be used for documentation or other test data.
+
+- As a part of the test step, we can run commands and execute scripts. Here, for arlon e2e tests, we execute the e2e_setup script as a part of the test step and then run the arlon commands specific to the test case.
+
+- Once we have created a test step case, we can also create a test assert for a given filename. The assert's filename should be the test step index followed by ```-assert.yaml``` e.g. In ```/testing/e2e/00-deploy``` we have ```02-assert.yaml```, which will run after ```02-deploy.yaml``` test step.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:33]:
```suggestion
- Once we have created a test step case, we can also create a test assert for a given filename. The assert's filename should be the test step index followed by `-assert.yaml` e.g. In `/testing/e2e/00-deploy` we have `02-assert.yaml`, which will run after `02-deploy.yaml` test step.
```

File: testing/e2e_docs.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+- In a test assert step, we look for the desired resource state. This is the state of the resource after the test step has finished. As a part of the test assert, we can add a timeout incase the resource that we are waiting on takes a long time to reach the desired state.
+
+- All the test steps and test asserts run in order and each must be successful for the test case to be considered successful.If any test step or test assertion fails then the test will fail.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:37]:
```suggestion
- All the test steps and test asserts run in order and each must be successful for the test case to be considered successful. If any test step or test assertion fails then the test will fail.
```

File: testing/e2e_docs.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+
+### Testbed Teardown
+
+- Currently, as a part of the ```/testing/e2e_setup_teardown.sh``` script, we delete the kind management cluster, the cloned workspace repository, bundles, profiles and cluster manifests present in this repository.
```
Comment (Rohitrajak1807) [inline testing/e2e_docs.md:42]:
```suggestion
- Currently, as a part of the `/testing/e2e_setup_teardown.sh` script, we delete the kind management cluster, the cloned workspace repository, bundles, profiles and cluster manifests present in this repository.
```

**Review by Rohitrajak1807 — CHANGES_REQUESTED** [review]

**Review by Rohitrajak1807 — APPROVED** [review]

File: testing/e2e_docs.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
+## Arlon e2e tests
```
Comment (cruizen) [inline testing/e2e_docs.md:1]: can you rename the file to `e2e_testing.md` or similar?

File: testing/e2e_docs.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+- This script runs at the end of every e2e test run regardless of the success or failure of the test for cleaning up any resources that might have been created as a part of the test.
+
+- Any additional resource that needs to be cleaned up post the e2e test should be added to this script.
+
+
+### e2e tests integration with Github Actions
+
+- Currently, we are running the arlon e2e tests on an ubuntu VM using Github Actions.
+
+- To run the arlon e2e tests, we need to add a new [make target](https://github.com/arlonproj/arlon/blob/main/Makefile#L154) for this new e2e test, which will then be executed in the [e2e test job](https://github.com/arlonproj/arlon/blob/main/.github/workflows/e2e.yaml) in Github Actions
```
Comment (cruizen) [inline testing/e2e_docs.md:53]:
```suggestion
- To invoke the arlon e2e tests, we are adding a new build target `make test-e2e` This command will execute the [e2e test job](https://github.com/arlonproj/arlon/blob/main/.github/workflows/e2e.yaml) using Github Actions
```

File: testing/e2e_docs.md (hunk 9 above)
Comment (ShaunakJoshi1407) [reply testing/e2e_docs.md]: Sure

File: testing/e2e_docs.md (hunk 10 above)
Comment (ShaunakJoshi1407) [reply testing/e2e_docs.md]: Done

**Review by cruizen — APPROVED** [review]
Approved, thanks.

**Pull request merged**